rotkehlchen/chain/ethereum/modules/aave/v3/balances.py

```diff
@@ -0,0 +1,129 @@
+import logging
+from collections import defaultdict
+from typing import TYPE_CHECKING
+
+from rotkehlchen.accounting.structures.balance import Balance
+from rotkehlchen.assets.asset import EvmToken
+from rotkehlchen.chain.balances import BlockchainBalances
+from rotkehlchen.chain.ethereum.interfaces.balances import ProtocolWithBalance
+from rotkehlchen.chain.ethereum.utils import token_normalized_value_decimals
+from rotkehlchen.chain.evm.contracts import EvmContract
+from rotkehlchen.chain.evm.decoding.aave.constants import CPT_AAVE_V3
+from rotkehlchen.chain.evm.types import string_to_evm_address
+from rotkehlchen.constants.misc import ZERO
+from rotkehlchen.constants.resolver import evm_address_to_identifier
+from rotkehlchen.errors.asset import UnknownAsset, WrongAssetType
+from rotkehlchen.errors.misc import RemoteError
+from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.inquirer import Inquirer
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import ChainID, EvmTokenKind, SupportedBlockchain
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.ethereum.interfaces.balances import BalancesType
+    from rotkehlchen.chain.ethereum.node_inquirer import EthereumInquirer
+    from rotkehlchen.db.dbhandler import DBHandler
+    from rotkehlchen.types import ChecksumEvmAddress
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class Aavev3Balances(ProtocolWithBalance):
+    def __init__(self, database: 'DBHandler', evm_inquirer: 'EthereumInquirer'):
+        super().__init__(
+            database=database,
+            evm_inquirer=evm_inquirer,
+            counterparty=CPT_AAVE_V3,
+            deposit_event_types={(HistoryEventType.DEPOSIT, HistoryEventSubType.DEPOSIT_ASSET)},
+        )
+        self.balances = BlockchainBalances(db=database).get(SupportedBlockchain.ETHEREUM)
+
+    def query_balances(self) -> 'BalancesType':
+        """
+        Query balances for Aave v3 open positions.
+        Docs: https://docs.aave.com/developers/guides/rates-guide#how-is-yield-accrued
+
+        The positions are fetched from the AToken contracts by calling the .balanceOf()
+        method on them. Then return the remaining balances that is not detected.
+        """
+        balances: BalancesType = defaultdict(lambda: defaultdict(Balance))
+        addresses_events = self.addresses_with_activity(
+            products=None,
+            event_types=self.deposit_event_types,
+        )
+        if len(addresses_events) == 0:
+            return balances
+
+        users_tokens: dict['ChecksumEvmAddress', set['ChecksumEvmAddress']] = defaultdict(set)  # unique tokens for each address  # noqa: E501
+        for user_address, events in addresses_events.items():
+            for event in events:
+                if event.address is not None:
+                    users_tokens[user_address].add(event.address)
```

Comment:
can't you make this a list comprehnsion? Also this starts to be an inneficient query that could be easily executed in sql uning a `select DISTINCT location_label, address from history events where type=? AND subtype=? AND counterparty=? AND location=?`

Reply:
This we can do in `ProtocolWithBalance:addresses_with_activity`. There is another TODO there, which will also help with this optimisation.